ledger/participant-integration-api/src/main/scala/platform/indexer/MeteringAggregator.scala

```diff
@@ -0,0 +1,180 @@
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+object MeteringAggregator {
+
+  class Owner(
+      meteringStore: MeteringStorageWriteBackend,
+      parameterStore: ParameterStorageBackend,
+      metrics: Metrics,
+      period: FiniteDuration = 6.minutes,
+  ) {
+
+    private[platform] def apply(
+        dispatcher: DbDispatcher
+    )(implicit loggingContext: LoggingContext): ResourceOwner[Unit] = {
+      val aggregator = new MeteringAggregator(meteringStore, parameterStore, metrics)
+      Await.ready(aggregator.initialize(dispatcher), 60.seconds)
+      ResourceOwner.forTimer(() => new Timer()).map { timer =>
+        timer.scheduleAtFixedRate(
+          new TimerTask {
+            override def run(): Unit = {
+              Await.ready(aggregator.run(dispatcher), 60.minutes)
+              ()
+            }
+          },
+          period.toMillis,
+          period.toMillis,
+        )
+      }
+    }
+  }
+
+  private def toTimestamp(dateTime: OffsetDateTime): Timestamp =
+    Timestamp.assertFromInstant(dateTime.toInstant)
+  private def toOffsetDateTime(timestamp: Timestamp): OffsetDateTime =
+    OffsetDateTime.ofInstant(timestamp.toInstant, ZoneOffset.UTC)
+
+}
+
+class MeteringAggregator(
+    meteringStore: MeteringStorageWriteBackend,
+    parameterStore: ParameterStorageBackend,
+    metrics: Metrics,
+    clock: () => Timestamp = () => Timestamp.now(),
+)(implicit loggingContext: LoggingContext) {
+
+  private implicit val ec: ExecutionContext = ExecutionContext.global
+
+  private val logger = ContextualizedLogger.get(getClass)
+
+  private[platform] def initialize(sqlExecutor: SqlExecutor): Future[Unit] = {
+    val initTimestamp = toOffsetDateTime(clock()).truncatedTo(ChronoUnit.HOURS).minusHours(1)
+    val initLedgerMeteringEnd = LedgerMeteringEnd(Offset.beforeBegin, toTimestamp(initTimestamp))
+    sqlExecutor.executeSql(metrics.daml.index.db.initializeMeteringAggregator) {
+      parameterStore.initializeLedgerMeteringEnd(initLedgerMeteringEnd)
+    }
+  }
+
+  private[platform] def run(sqlExecutor: SqlExecutor): Future[Unit] = {
+
+    val future = sqlExecutor.executeSql(metrics.daml.index.db.meteringAggregator) { conn =>
+      val nowUtcTime = toOffsetDateTime(clock())
+      val lastLedgerMeteringEnd = getLedgerMeteringEnd(conn)
+      val startUtcTime: OffsetDateTime = toOffsetDateTime(lastLedgerMeteringEnd.timestamp)
+      val endUtcTime = startUtcTime.plusHours(1)
+
+      if (nowUtcTime.isAfter(endUtcTime)) {
+
+        val toEndTime = toTimestamp(endUtcTime)
+        val ingestedLedgerEnd = parameterStore.ledgerEnd(conn).lastOffset
+        val maybeMaxOffset =
+          meteringStore.transactionMeteringMaxOffset(lastLedgerMeteringEnd.offset, toEndTime)(conn)
+
+        val (
+          periodIngested, // This is true if the time period is closed fully ingested
+          hasMetering, // This is true if there are transaction_metering records
```

Comment:
Would you consider removing hasMetering altogether, and passing directly the maybeMaxOffset down?
IMHO it would be much better readable if the decision for the hasMetering logic is next to the conditional branching in aggregate (this applies to toOffsetEnd: that one might be more natural to deduce as we would do the branching in aggregate)

Reply:
The way the code started the three fields were all derived at different times based on a function of maybeMaxOffset and it confused me and was difficult to read.

Now we have a situation where all the maybeMaxOffset switching is done in one place (albeit slightly complicated).

This has the great upside that from this point all the values are explicit.